// @fluentui/react-positioning/lib/useSafeZoneArea.js
import { useFluent_unstable } from '@fluentui/react-shared-contexts';
import { useEventCallback, useMergedRefs } from '@fluentui/react-utilities';
import * as React from 'react';
import { createSafeZoneAreaStateStore, SafeZoneArea } from './SafeZoneArea';
import { parseFloatingUIPlacement } from './utils';
export function useSafeZoneArea({ debug = false, disabled = false, onSafeZoneEnter, onSafeZoneMove, onSafeZoneLeave, onSafeZoneTimeout, timeout = 1500 } = {}) {
    const [stateStore] = React.useState(createSafeZoneAreaStateStore);
    const { targetDocument } = useFluent_unstable();
    const safeZoneAreaRef = React.useRef(null);
    const containerRef = React.useRef(null);
    const targetRef = React.useRef(null);
    const timeoutIdRef = React.useRef(null);
    const mouseMoveIdRef = React.useRef(null);
    const mouseCoordinatesRef = React.useRef({
        x: 0,
        y: 0
    });
    const containerListenerRef = React.useMemo(()=>{
        if (disabled) {
            return ()=>{
            // do nothing
            };
        }
        let containerEl = null;
        function onContainerMouseEnter() {
            const targetWindow = targetDocument === null || targetDocument === void 0 ? void 0 : targetDocument.defaultView;
            if (!targetWindow) {
                return;
            }
            if (timeoutIdRef.current) {
                targetWindow.clearTimeout(timeoutIdRef.current);
                timeoutIdRef.current = null;
            }
            stateStore.toggleActive(false);
        }
        return (el)=>{
            if (el === null) {
                containerEl === null || containerEl === void 0 ? void 0 : containerEl.removeEventListener('mouseenter', onContainerMouseEnter);
            }
            containerEl = el;
            el === null || el === void 0 ? void 0 : el.addEventListener('mouseenter', onContainerMouseEnter);
        };
    }, [
        disabled,
        stateStore,
        targetDocument
    ]);
    const targetListenerRef = React.useMemo(()=>{
        if (disabled) {
            return ()=>{
            // do nothing
            };
        }
        let targetEl = null;
        function onTargetMouseMove(e) {
            mouseCoordinatesRef.current = {
                x: e.clientX,
                y: e.clientY
            };
            if (timeoutIdRef.current) {
                var _targetDocument_defaultView;
                targetDocument === null || targetDocument === void 0 ? void 0 : (_targetDocument_defaultView = targetDocument.defaultView) === null || _targetDocument_defaultView === void 0 ? void 0 : _targetDocument_defaultView.clearTimeout(timeoutIdRef.current);
                timeoutIdRef.current = null;
            }
            if (!stateStore.isActive()) {
                stateStore.toggleActive(true);
            }
        }
        return (el)=>{
            if (el === null) {
                const targetWindow = targetDocument === null || targetDocument === void 0 ? void 0 : targetDocument.defaultView;
                if (targetWindow) {
                    if (mouseMoveIdRef.current) {
                        targetWindow.cancelAnimationFrame(mouseMoveIdRef.current);
                        mouseMoveIdRef.current = null;
                    }
                    if (timeoutIdRef.current) {
                        targetWindow.clearTimeout(timeoutIdRef.current);
                        timeoutIdRef.current = null;
                    }
                }
                targetEl === null || targetEl === void 0 ? void 0 : targetEl.removeEventListener('mousemove', onTargetMouseMove);
            }
            targetEl = el;
            el === null || el === void 0 ? void 0 : el.addEventListener('mousemove', onTargetMouseMove);
        };
    }, [
        disabled,
        stateStore,
        targetDocument
    ]);
    const onSvgMouseEnter = useEventCallback((e)=>{
        onSafeZoneEnter === null || onSafeZoneEnter === void 0 ? void 0 : onSafeZoneEnter(e);
        const targetWindow = targetDocument === null || targetDocument === void 0 ? void 0 : targetDocument.defaultView;
        if (!targetWindow) {
            return;
        }
        if (timeoutIdRef.current) {
            targetWindow.clearTimeout(timeoutIdRef.current);
            timeoutIdRef.current = null;
        }
        // React 17 still uses pooled synthetic events
        e.persist();
        timeoutIdRef.current = targetWindow.setTimeout(()=>{
            stateStore.toggleActive(false);
            onSafeZoneTimeout === null || onSafeZoneTimeout === void 0 ? void 0 : onSafeZoneTimeout();
        }, timeout);
    });
    const onSvgMouseMove = useEventCallback((e)=>{
        onSafeZoneMove === null || onSafeZoneMove === void 0 ? void 0 : onSafeZoneMove(e);
    });
    const onSvgMouseLeave = useEventCallback((e)=>{
        onSafeZoneLeave === null || onSafeZoneLeave === void 0 ? void 0 : onSafeZoneLeave(e);
    });
    React.useEffect(()=>{
        return stateStore.subscribe((isActive)=>{
            if (isActive) {
                function updateSVGs() {
                    const containerEl = containerRef.current;
                    const targetEl = targetRef.current;
                    const targetWindow = targetDocument === null || targetDocument === void 0 ? void 0 : targetDocument.defaultView;
                    if (containerEl && targetEl) {
                        var _safeZoneAreaRef_current;
                        (_safeZoneAreaRef_current = safeZoneAreaRef.current) === null || _safeZoneAreaRef_current === void 0 ? void 0 : _safeZoneAreaRef_current.updateSVG({
                            containerPlacementSide: parseFloatingUIPlacement(containerEl.dataset.popperPlacement).side,
                            containerRect: containerEl.getBoundingClientRect(),
                            mouseCoordinates: mouseCoordinatesRef.current,
                            targetRect: targetEl.getBoundingClientRect()
                        });
                    }
                    if (targetWindow) {
                        mouseMoveIdRef.current = targetWindow.requestAnimationFrame(updateSVGs);
                    }
                }
                updateSVGs();
                return;
            }
            if (mouseMoveIdRef.current) {
                var _targetDocument_defaultView;
                targetDocument === null || targetDocument === void 0 ? void 0 : (_targetDocument_defaultView = targetDocument.defaultView) === null || _targetDocument_defaultView === void 0 ? void 0 : _targetDocument_defaultView.cancelAnimationFrame(mouseMoveIdRef.current);
                mouseMoveIdRef.current = null;
            }
        });
    }, [
        stateStore,
        targetDocument
    ]);
    return {
        containerRef: useMergedRefs(containerRef, containerListenerRef),
        targetRef: useMergedRefs(targetRef, targetListenerRef),
        elementToRender: React.useMemo(()=>disabled ? null : /*#__PURE__*/ React.createElement(SafeZoneArea, {
                debug: debug,
                onMouseEnter: onSvgMouseEnter,
                onMouseMove: onSvgMouseMove,
                onMouseLeave: onSvgMouseLeave,
                imperativeRef: safeZoneAreaRef,
                stateStore: stateStore
            }), [
            disabled,
            debug,
            onSvgMouseEnter,
            onSvgMouseMove,
            onSvgMouseLeave,
            stateStore
        ])
    };
}
